### PR TITLE
Proposal for a convention for a new importer function

### PR DESCRIPTION
**Pull request by catalan-adobe**
Proposal for a convention for a new importer function to execute javascript after page load.

This could be used to wait for specific DOM element to be present before execution the transform function

Using "convention" approach, this feature can be kept only for advanced/complex use cases without need to expose (for example) a new field in the UI.
It could be documented in the guidelines.

⚠️ I would re-create a PR to take in account #101

## bacom example

https://business.adobe.com/au/resources/reports/3partstrategyforuppingyourgame.html

The page loads an external form which can take few seconds to be added to the DOM.
Rather than adding a fixed waiting time which would greatly affect bulk imports, this proposal would allow following mechanism in import.js:

1. At the top of import.js, add util function to wait for a DOM element to exist
```
[...]
async function awaitElement(selector, main, timeout = 10000) {
  return new Promise((resolve, reject) => {
    const timeWas = new Date();
    const wait = setInterval(function() {

      console.log('poll check', main.querySelector(selector));
      if (main.querySelector(selector)) {
        console.log("resolved after", new Date() - timeWas, "ms");
        clearInterval(wait);
        resolve();
      } else if (new Date() - timeWas > timeout) { // Timeout
        console.log("rejected after", new Date() - timeWas, "ms");
        clearInterval(wait);
        reject();
      }
    }, 200);
  });
}
[...]
```

2. In `export default {`, add following "convention" function `postLoadScript`
```
[...]
export default {
  postLoadScript: (async (document) => {
    try {
      await awaitElement('.faas-form-settings', main, 2500);
      await awaitElement('.faas-form', main, 2500);
    } catch (e) {
      console.error('no form found!', url);
    }
  }),
[...]
transformDOM: ({
[...]
```

### CODE REVIEW

**Comment by kptdobe** [comment]
Also, I feel like this wait operation is the feature want to have, not a generic entry point into something as a user I would know what to do with.

File: js/import/import.ui.js [hunk 1]
```diff
@@ -339,6 +339,11 @@ const attachListeners = () => {
                       includeDocx,
                       params: { originalURL },
                     });
+
+                    if (config.importer.projectTransform.onLoad) {
```
Comment (kptdobe) [inline js/import/import.ui.js:343]:
```suggestion
                    if (config.importer.projectTransform?.onLoad) {
```

File: js/import/import.ui.js [hunk 2]
```diff
@@ -339,6 +339,11 @@ const attachListeners = () => {
...
+                      await config.importer.projectTransform.onLoad(frame.contentDocument);
```
Comment (kptdobe) [inline js/import/import.ui.js:344]:
```suggestion
                      await config.importer.projectTransform.onLoad({ 
                          document: frame.contentDocument,
                          url: replacedURL,
                          params: { originalURL },                          
                      });
```

Comment (kptdobe) [reply js/import/import.ui.js:344]: To get the similar set of parameters than other methods.

**Review by kptdobe — CHANGES_REQUESTED** [review]

File: js/import/import.ui.js (hunk 2 above)
Comment (catalan-adobe) [reply js/import/import.ui.js]: fixed in https://github.com/adobe/helix-importer-ui/pull/102/commits/43d9244e81b32ecd8d1046befe07860336136ba1

**Comment by catalan-adobe** [comment]
@kptdobe, ready for another round of review :) Feedback should be adressed:
* [d061bcc](https://github.com/adobe/helix-importer-ui/pull/102/commits/d061bcc127ef027c93f8cbbf9f194620ffe51e29) merged main, no more conflicts
* [a1dab3d](https://github.com/adobe/helix-importer-ui/pull/102/commits/a1dab3d4c9a3a07d9590e5057e3805500f65127c) updated helix-importer dependency and executed `npm run build:prod`
* [43d9244](https://github.com/adobe/helix-importer-ui/pull/102/commits/43d9244e81b32ecd8d1046befe07860336136ba1) refactored `onLoad` based on latest feedback

File: js/shared/pollimporter.js [hunk 3]
```diff
@@ -12,6 +12,8 @@
 /* global WebImporter */
 
 export default class PollImporter {
+  #onLoad = async () => {};
```
Comment (kptdobe) [inline js/shared/pollimporter.js:15]: useless no ?

**Review by kptdobe — CHANGES_REQUESTED** [review]
Please also remove [js/dist/helix-importer.js](https://github.com/adobe/helix-importer-ui/pull/102/files#diff-9b9a5e97d6d559d3f229bc2114e0da7fbc4de94d1a68dcf0351b3b2aaac46046) from PR

File: js/shared/pollimporter.js (hunk 3 above)
Comment (catalan-adobe) [reply js/shared/pollimporter.js]: indeed!

File: package.json [hunk 4]
```diff
@@ -14,7 +14,7 @@
     "semantic-release": "semantic-release"
   },
   "dependencies": {
-    "@adobe/helix-importer": "2.7.0",
+    "@adobe/helix-importer": "2.8.0",
```
[the remark this answers is not in the document]
Comment (catalan-adobe) [reply package.json]: done.

File: js/shared/pollimporter.js [hunk 5]
```diff
@@ -81,6 +81,29 @@ export default class PollImporter {
...
 
+  async onLoad({ url, document, params }) {
+    if (this.projectTransform && this.projectTransform.onLoad) {
+      try {
+        await this.projectTransform.onLoad({
+          url,
+          document,
+          params,
+        });
+        return true;
```
Comment (kptdobe) [inline js/shared/pollimporter.js:92]:
```suggestion
```
This `return true` is useless.

**Review by kptdobe — APPROVED** [review]

File: js/shared/pollimporter.js (hunk 5 above)
Comment (catalan-adobe) [reply js/shared/pollimporter.js]: Good catch!

**Pull request merged**